# Введение

В задачах 3, 4 были реализованы 2 алгоритма для выполнения регулярных запросов к графам. Первый алгоритм представляет из себя пересечение двух автоматов при помощи произведения Кронокера и последующее транзиктивное замыкание его матрицы смежности, то есть поиск путей сразу от всех ко всем, второй –– представляет синхронный обход в ширину двух автоматов с возможностью обходить по отдельности так, если бы фронт состоял из одной стартовой вершины и сразу из всего стартового множества, то есть фронт –– множество всех стартовые вершин.

*алгоритмом номер 2(фронт –– множество) будем называть BfsMultipleSource*

*алгоритмом номер 2(фронт –– 1 стартовая вершина) будем называть BfsSingleSource*

## Целью данной работы является:
* Определить какое представление разреженных матриц и векторов лучше подходит для каждого алгоритма
* Определить начиная с какого размера стартового множества выгоднее решать задачу алгоритмом номер 1 и алгоритмом номер 2
* Определить на сколько решение BfsMultipleSource быстрее решения BfsSingleSource при одинаковых начальных условиях



# Описание исследуемых решений

На вход поступает регулярное выражение, граф, и множества стартовых и конечных вершин графа (состояний). Регулярное выражение и граф переводятся в соответсвующие конечные автоматы.

## Алгоритм 1 (произведение Кронокера)

У получившихся автоматов находятся булевы матрицы смежности. Далее получаем пересечение данных автоматов в виде автомата с соответсвующими булевыми матрицами смежности, полученными от произведения Кронокера двух соответсвующих матриц смежности. После, с помощью транзиктивного замыкания (умножения матрицы смежности саму на себя) находим пути достижимости от всех вершин ко всем. После фильтруем результат транзиктивного замыкания, операясь на информацию о стартовых и конечных состояниях пересечения автоматов. Итог –– множество пар достижимости от стартовой до конечной вершины.

## Алгоритм 2 (BfsMultipleSource)

Получившиеся автоматы синхронно обходим в ширину до тех пор пока можно ходить одновременно по двум автоматам по одинаковым ребрам. Синхронный обход производится путем умножения матрицы фронта на матрицу прямой суммы булевых матриц двух автоматов. Матрица фронта ––  матрица K x K+N (K –– кол-во состояний в 1-ом автомате, N –– кол-во состояний в 2-ом автомате). Первый блок KxK –– диагональная матрица, где на диагонали стоит True с индексом ixi, если индекс i –– индекс стартовой вершины в 1-ом автомате. Второй блок правее от блока KxK, его размерность –– KxN. True с индексом ixj, если i –– стартовое состояние автомата 1 и j –– стартовое состояние автомата 2.

## Алгоритм 2 (BfsSingleSource)

Получившиеся автоматы синхронно обходим в ширину до тех пор пока можно ходить одновременно по двум автоматам по одинаковым ребрам. Синхронный обход производится путем умножения матрицы фронта на матрицу прямой суммы булевых матриц двух автоматов. Матрица фронта ––  матрица K*K x K+N (K –– кол-во состояний в 1-ом автомате, N –– кол-во состояний в 2-ом автомате). Левые блоки KxK –– матрицы, где на диагонали стоят True с индексом ixi, если индекс i –– индекс стартовой вершины в 1-ом автомате (такое True всего одно для каждой матрицы KxK). Вторые блоки правее от блока KxK, его размерность –– KxN. True с индексом ixj, если i –– стартовое состояние автомата 1 и j –– стартовое состояние автомата 2.

# Описание набора данных для экспериментов

Для проведения экспериментов были использованы графы из [датасета](https://jetbrains-research.github.io/CFPQ_Data/reference/dataset/generated/cfpq_data.dataset.DATASET.html), а именно:

1. **skos: 144 вершины**, **252 ребра**, labels: {'domain', 'unionOf', 'subClassOf', 'title', 'seeAlso', 'rest', 'example', 'creator', 'scopeNote', 'definition', 'type', 'contributor', 'isDefinedBy', 'first', 'inverseOf', 'range', 'subPropertyOf', 'disjointWith', 'comment', 'label', 'description'}

2. **wc: 332 вершины, 269 ребер**, labels: {'a', 'd'}

3. **funding: 778 вершин, 1086 ребер**, labels: {'imports', 'domain', 'unionOf', 'subClassOf', 'date', 'title', 'seeAlso', 'versionInfo', 'rest', 'rights', 'creator', 'type', 'contributor', 'first', 'inverseOf', 'range', 'subPropertyOf', 'comment', 'label', 'description'}

4. **bzip: 632 вершины, 556 ребер**, labels: {'a', 'd'}

5. **gzip: 2687 вершин**, **2293 ребра**, labels: {'a', 'd'}

6. **pathways: 6238 вершин**, **12363 ребра**, labels: {'imports', 'narrower', 'subClassOf', 'type', 'label'}

Графы выбирались по увелечению кол-ва ребер и вершин, чтобы "Определить начиная с какого размера стартового множества выгоднее решать задачу алгоритмом номер 1 и алгоритмом номер 2" и чтобы "Определить на сколько решение BfsMultipleSource быстрее решения BfsSingleSource при одинаковых начальных условиях". Так же выбирались графы с разными (по размеру) алфавитами лейблов.

Графы выбирались не совсем большими, чтобы эксперимент работал приемлимое время на слабом железе. 

В качестве регулярных выражений для экспериментов были выбраны 3 вида:

1. Регулярное выражение, допускающее весь графа
2. Регулярное выражение, допускающее один символ из алфавита лейблов
3. Регулярное выражение, содержащее все основные конструкции регулярных выражений


# Описание эксперимента

## Оборудование

* Процессор: [1,1 GHz 4‑ядерный процессор Intel Core i5](https://ark.intel.com/content/www/ru/ru/ark/products/196603/intel-core-i51035g1-processor-6m-cache-up-to-3-60-ghz.html)
* Оперативная память: 8 ГБ 3733 MHz LPDDR4X

## Что и как замерялось, как эти измерения должны помочь ответить на поставленные вопросы

Измерения проводились для кажого графа и соответсвующего регулярного выражения 20 раз, считалось средняя время поиска достижимости и среднее квадратичное отклонение этого времени.

Измерения должны ответить на вопросы: 
* Начиная с какого размера стартового множества выгоднее решать задачу алгоритмом номер 1 и алгоритмом номер 2
* На сколько решение BfsMultipleSource быстрее решения BfsSingleSource при одинаковых начальных условиях

Графы будут браться в порядке увелечения кол-ва вершин и ребер, после чего можно будет примерно определить начиная с какого размера стартового множества выгоднее решать задачу алгоритмом номер 1 и алгоритмом номер 2. Так же можно будет более точно ответить на сколько решение BfsMultipleSource быстрее решения BfsSingleSource при одинаковых начальных условиях.


In [51]:
import sys
import os
path = os.getcwd()[:-11]
if path not in sys.path: sys.path.append(os.getcwd()[:-11])
    
import random
import statistics

from timeit import timeit
from networkx import MultiDiGraph
from pyformlang.regular_expression import Regex
from cfpq_data import generate_multiple_source_percent

from project.graph_utilities import get_graph
from project.rpq import rpq
from project.rpq import MethodRpq
from project.rpq import MethodBfs
from project.rpq import MethodTensor

import logging
logging.getLogger().disabled = True
# now logger will not log to console.

graph_names = ["skos", "funding", "wc", "bzip", "gzip", "pathways"] 
graphs = {graph_name: get_graph(graph_name) for graph_name in graph_names}
execution_times_number = 20

float_fixed = 5
def to_fixed(numObj, digits=0):
    return f"{numObj:.{digits}f}"

def generate_regexes(graph: MultiDiGraph) -> tuple[str, str, str]:
    graph_lables = list({el["label"] for (_, _, el) in graph.edges(data=True)})

    regex1 = f"({'+'.join(graph_lables)})*"

    regex2 = f"{random.choice(graph_lables)}"

    def regex3():
        a, b = random.choices(graph_lables, k=2)
        while b == a:
            b = random.choice(graph_lables)
        return f'{a}.{a}*.({b}+{a})'

    return regex1, regex2, regex3()

def repeat(stmt) -> str:
    times = []
    for i in range(execution_times_number):
        times.append(timeit(stmt=stmt, number=1))
    return f"{to_fixed(statistics.mean(times), float_fixed)} +- {to_fixed(statistics.stdev(times), float_fixed)}"

experiment_data = []
experiment_data_matrix_test = []
experiment_regexes = {
    graph_name: generate_regexes(graphs.get(graph_name))
    for graph_name in graphs.keys()
}

def fill_experiment_data_matrix_test(percent: float, seed: int):
    for graph_name in graph_names:
        if graph_name in ["gzip", "pathways"]:
            if not (0.1 <= percent <= 0.3): continue
        elif (0.1 <= percent <= 0.3): continue
        print(f"{graph_name} started")
        graph = graphs.get(graph_name)
        tuple_experiment_regexes = experiment_regexes.get(graph_name)
        regex = tuple_experiment_regexes[0]
        algo1csr = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed)))
        algo1csc = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), csr=False))
        experiment_data_matrix_test.append([graph_name, "tensor", "(Union_all)*", f"{percent}%", "csr", algo1csr])
        experiment_data_matrix_test.append([graph_name, "tensor", "(Union_all)*", f"{percent}%", "csc", algo1csc])
        print(f"{graph_name} completed part 1")
        algo2csr = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=False))))
        algo2csc = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=False)), csr=False))
        experiment_data_matrix_test.append([graph_name, "bfs_ms", "(Union_all)*", f"{percent}%", "csr", algo2csr])
        experiment_data_matrix_test.append([graph_name, "bfs_ms", "(Union_all)*", f"{percent}%", "csc", algo2csc])
        print(f"{graph_name} completed part 2")
        algo3csr = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=True))))
        algo3csc = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=True)), csr=False))
        experiment_data_matrix_test.append([graph_name, "bfs_ss", "(Union_all)*", f"{percent}%", "csr", algo3csr])
        experiment_data_matrix_test.append([graph_name, "bfs_ss", "(Union_all)*", f"{percent}%", "csc", algo3csc])
        print(f"{graph_name} completed fully")

def fill_experiment_data(percent: float, seed: int):
    for graph_name in graph_names:
        if graph_name in ["gzip", "pathways"]:
            if not (0.1 <= percent <= 0.3): continue
        elif (0.1 <= percent <= 0.3): continue
        print(f"{graph_name} started")
        graph = graphs.get(graph_name)
        tuple_experiment_regexes = experiment_regexes.get(graph_name)
        regex = tuple_experiment_regexes[0]
        algo1 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed)))
        algo2 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=False))))
        algo3 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=True))))
        experiment_data.append([graph_name, "tensor", "(Union_all)*", f"{percent}%", algo1])
        experiment_data.append([graph_name, "bfs_ms", "(Union_all)*", f"{percent}%", algo2])
        experiment_data.append([graph_name, "bfs_ss", "(Union_all)*", f"{percent}%", algo3])

        regex = tuple_experiment_regexes[1]
        algo1 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed)))
        algo2 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=False))))
        algo3 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=True))))
        experiment_data.append([graph_name, "tensor", f"{regex}", f"{percent}%", algo1])
        experiment_data.append([graph_name, "bfs_ms", f"{regex}", f"{percent}%", algo2])
        experiment_data.append([graph_name, "bfs_ss", f"{regex}", f"{percent}%", algo3])

        regex = tuple_experiment_regexes[2]
        algo1 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed)))
        algo2 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=False))))
        algo3 = repeat(lambda: rpq(regex, graph, start_states = generate_multiple_source_percent(graph, percent, seed=seed), method_rpq=MethodRpq(MethodBfs(is_single_source=True))))
        experiment_data.append([graph_name, "tensor", f"{regex}", f"{percent}%", algo1])
        experiment_data.append([graph_name, "bfs_ms", f"{regex}", f"{percent}%", algo2])
        experiment_data.append([graph_name, "bfs_ss", f"{regex}", f"{percent}%", algo3])
        print(f"{graph_name} completed fully")
print("Load ended")


Load ended


Будем использовать prettytable для того, чтобы создавать готовые таблицы:

In [52]:
import prettytable
from prettytable import PrettyTable

def print_table(header: list[str], rows: list[list[any]]):
    table = PrettyTable()
    table.set_style(prettytable.MARKDOWN)
    table.field_names = header
    for row in rows:
        table.add_row(row)
    table.max_width = 33
    print(table)
    

# Эксперимент

## Какое представление разреженных матриц лучше подходит для каждой реализации алгоритма?

Так как во всех 3-х алгоритмах в качестве основной части вычислений выполняется произведение матриц, было решено использовать [csr](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix) и [csc](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix) матрицы из [scipy.sparse](https://docs.scipy.org/doc/scipy/reference/sparse.html), которые быстро выполняют эту операцию. Также при проведении экспериментов, возьмем только регулярное выражение допускающее все возможные переходы в графе, чтобы увеличить кол-во умножений матриц в алгоритмах.

Выполняем эксперимент:

In [53]:
seed = 31
fill_experiment_data_matrix_test(0.2, seed)
fill_experiment_data_matrix_test(1.0, seed)
fill_experiment_data_matrix_test(10.0, seed)
fill_experiment_data_matrix_test(30.0, seed)
print("experiment ended fully")

gzip started
gzip completed part 1
gzip completed part 2
gzip completed fully
pathways started
pathways completed part 1
pathways completed part 2
pathways completed fully
skos started
skos completed part 1
skos completed part 2
skos completed fully
funding started
funding completed part 1
funding completed part 2
funding completed fully
wc started
wc completed part 1
wc completed part 2
wc completed fully
bzip started
bzip completed part 1
bzip completed part 2
bzip completed fully
skos started
skos completed part 1
skos completed part 2
skos completed fully
funding started
funding completed part 1
funding completed part 2
funding completed fully
wc started
wc completed part 1
wc completed part 2
wc completed fully
bzip started
bzip completed part 1
bzip completed part 2
bzip completed fully
skos started
skos completed part 1
skos completed part 2
skos completed fully
funding started
funding completed part 1
funding completed part 2
funding completed fully
wc started
wc completed part

Печатаем таблицу с данными эксперимента:

In [59]:
rows = []
for i in range(0, len(experiment_data_matrix_test), 6):
    rows.append([experiment_data_matrix_test[i][0], experiment_data_matrix_test[i][2], experiment_data_matrix_test[i][3], experiment_data_matrix_test[i][5], experiment_data_matrix_test[i+1][5], experiment_data_matrix_test[i+2][5], experiment_data_matrix_test[i+3][5], experiment_data_matrix_test[i+4][5], experiment_data_matrix_test[i+5][5]])
print_table(["graph_name", "regex", "percent_starts", "time_csr_tensor", "time_csc_tensor", "time_csr_bfs_ms", "time_csc_bfs_ms", "time_csr_bfs_ss", "time_csc_bfs_ss"], rows)

| graph_name |    regex     | percent_starts |  time_csr_tensor   |  time_csc_tensor   |  time_csr_bfs_ms   |  time_csc_bfs_ms   |   time_csr_bfs_ss   |   time_csc_bfs_ss   |
|:----------:|:------------:|:--------------:|:------------------:|:------------------:|:------------------:|:------------------:|:-------------------:|:-------------------:|
|    gzip    | (Union_all)* |      0.2%      | 0.07476 +- 0.00686 | 0.06949 +- 0.00533 | 0.11448 +- 0.00886 | 0.11210 +- 0.00776 |  0.11901 +- 0.00726 |  0.11717 +- 0.00529 |
|  pathways  | (Union_all)* |      0.2%      | 2.64071 +- 0.09793 | 3.45211 +- 0.84291 | 0.88362 +- 0.19061 | 0.77802 +- 0.01801 |  1.29093 +- 0.01498 |  1.27261 +- 0.02147 |
|    skos    | (Union_all)* |      1.0%      | 0.13309 +- 0.00962 | 0.12780 +- 0.00563 | 0.13246 +- 0.00601 | 0.13456 +- 0.00935 |  0.13456 +- 0.00797 |  0.13627 +- 0.01075 |
|  funding   | (Union_all)* |      1.0%      | 0.13980 +- 0.00543 | 0.13960 +- 0.00914 | 0.15863 +- 0.00593 | 0.15454 +- 0.00

    Таблица:
| graph_name |    regex     | percent_starts |  time_csr_tensor   |  time_csc_tensor   |  time_csr_bfs_ms   |  time_csc_bfs_ms   |   time_csr_bfs_ss   |   time_csc_bfs_ss   |
|:----------:|:------------:|:--------------:|:------------------:|:------------------:|:------------------:|:------------------:|:-------------------:|:-------------------:|
|    gzip    | (Union_all)* |      0.2%      | 0.07476 +- 0.00686 | 0.06949 +- 0.00533 | 0.11448 +- 0.00886 | 0.11210 +- 0.00776 |  0.11901 +- 0.00726 |  0.11717 +- 0.00529 |
|  pathways  | (Union_all)* |      0.2%      | 2.64071 +- 0.09793 | 3.45211 +- 0.84291 | 0.88362 +- 0.19061 | 0.77802 +- 0.01801 |  1.29093 +- 0.01498 |  1.27261 +- 0.02147 |
|    skos    | (Union_all)* |      1.0%      | 0.13309 +- 0.00962 | 0.12780 +- 0.00563 | 0.13246 +- 0.00601 | 0.13456 +- 0.00935 |  0.13456 +- 0.00797 |  0.13627 +- 0.01075 |
|  funding   | (Union_all)* |      1.0%      | 0.13980 +- 0.00543 | 0.13960 +- 0.00914 | 0.15863 +- 0.00593 | 0.15454 +- 0.00427 |  0.19044 +- 0.00922 |  0.18963 +- 0.00610 |
|     wc     | (Union_all)* |      1.0%      | 0.01228 +- 0.00087 | 0.01198 +- 0.00072 | 0.01490 +- 0.00076 | 0.01450 +- 0.00112 |  0.01596 +- 0.00136 |  0.01612 +- 0.00090 |
|    bzip    | (Union_all)* |      1.0%      | 0.02070 +- 0.00177 | 0.02234 +- 0.00227 | 0.03179 +- 0.00177 | 0.03249 +- 0.00217 |  0.04388 +- 0.00522 |  0.04118 +- 0.00263 |
|    skos    | (Union_all)* |     10.0%      | 0.12876 +- 0.00719 | 0.12995 +- 0.00631 | 0.21152 +- 0.01088 | 0.20985 +- 0.00589 |  0.43895 +- 0.01518 |  0.43562 +- 0.01226 |
|  funding   | (Union_all)* |     10.0%      | 0.13604 +- 0.00421 | 0.14050 +- 0.00974 | 0.21181 +- 0.00713 | 0.21442 +- 0.01079 |  3.09361 +- 0.02868 |  3.05965 +- 0.03670 |
|     wc     | (Union_all)* |     10.0%      | 0.01256 +- 0.00200 | 0.01286 +- 0.00224 | 0.03313 +- 0.00384 | 0.03206 +- 0.00189 |  0.20376 +- 0.00672 |  0.20077 +- 0.00649 |
|    bzip    | (Union_all)* |     10.0%      | 0.02118 +- 0.00118 | 0.02108 +- 0.00155 | 0.04040 +- 0.00208 | 0.04062 +- 0.00239 |  0.45972 +- 0.04087 |  0.44255 +- 0.00924 |
|    skos    | (Union_all)* |     30.0%      | 0.12394 +- 0.00512 | 0.12602 +- 0.00645 | 0.20998 +- 0.00793 | 0.21314 +- 0.00614 |  1.65030 +- 0.01301 |  1.64650 +- 0.01569 |
|  funding   | (Union_all)* |     30.0%      | 0.13464 +- 0.00750 | 0.13413 +- 0.00647 | 0.22493 +- 0.00519 | 0.22512 +- 0.00660 | 24.29073 +- 0.22980 | 26.87881 +- 3.26511 |
|     wc     | (Union_all)* |     30.0%      | 0.03071 +- 0.00645 | 0.02042 +- 0.00460 | 0.06910 +- 0.01544 | 0.06250 +- 0.01232 |  1.35669 +- 0.30353 |  1.31543 +- 0.16028 |
|    bzip    | (Union_all)* |     30.0%      | 0.02546 +- 0.00268 | 0.02519 +- 0.00253 | 0.06502 +- 0.00389 | 0.06552 +- 0.00473 |  6.41268 +- 0.07095 |  6.35344 +- 0.03176 |

### По данным эксперимента можно сделать выводы:
Нет разницы в использовании этих двух матриц. Результаты эксперимента вполне ожидаемы, так как по документации csr и csc должны отличаться только в скорости slicing. У csr быстрый row slicing и медленный column slicing. У csc быстрый column slicing и медленный row slicing. И так как slicing в алгоритмах не использовался (брались только nonezero элементы всей матрицы сразу), поэтому и разницы нет в использовании csr и csc. 

## Начиная с какого размера стартового множества выгоднее решать задачу алгоритмом номер 1 и алгоритмом номер 2 ?

Для того чтобы определить, начиная с какого размера стартового множества выгоднее решать задачу алгоритмом номер 2 можно оперировать процентом размера самого множества вершин в графе, где процент –– "кол-во" стартовых вершин относительно общего кол-ва вершин в графе. Логично было бы предположить, что чем больше граф, но в то же время процент стартовых вершин очень маленький, то алгоритм 2 должен быть более выгодным, чем алгоритм 1. Поэтому были взяты более большие графы как: gzip и pathways, а процент стартовых вершин был взять 0,2%. Для всех остальных графов проценты стартового множества будут: 1.0%, 10.0%, 30.0%.


Выполняем эксперимент:

In [60]:
seed = 31
float_fixed = 5
fill_experiment_data(0.2, seed)
fill_experiment_data(1.0, seed)
fill_experiment_data(10.0, seed)
fill_experiment_data(30.0, seed)
print("experiment ended fully")

gzip started
gzip completed fully
pathways started
pathways completed fully
skos started
skos completed fully
funding started
funding completed fully
wc started
wc completed fully
bzip started
bzip completed fully
skos started
skos completed fully
funding started
funding completed fully
wc started
wc completed fully
bzip started
bzip completed fully
skos started
skos completed fully
funding started
funding completed fully
wc started
wc completed fully
bzip started
bzip completed fully
experiment ended fully


Печатаем таблицу с данными эксперимента:

In [61]:
rows = []
for i in range(0, len(experiment_data), 3):
    rows.append([experiment_data[i][0], experiment_data[i][2], experiment_data[i][3], experiment_data[i][4], experiment_data[i+1][4], experiment_data[i+2][4]])
print_table(["graph_name", "regex", "percent_starts", "time_tensor", "time_bfs_ms", "time_bfs_ss"], rows)

| graph_name |               regex               | percent_starts |    time_tensor     |    time_bfs_ms     |     time_bfs_ss     |
|:----------:|:---------------------------------:|:--------------:|:------------------:|:------------------:|:-------------------:|
|    gzip    |            (Union_all)*           |      0.2%      | 0.06600 +- 0.00919 | 0.11013 +- 0.00951 |  0.11492 +- 0.00490 |
|    gzip    |                 a                 |      0.2%      | 0.05386 +- 0.00303 | 0.04389 +- 0.00089 |  0.04527 +- 0.00237 |
|    gzip    |             a.a*.(d+a)            |      0.2%      | 0.10893 +- 0.00453 | 0.10059 +- 0.00840 |  0.11142 +- 0.00589 |
|  pathways  |            (Union_all)*           |      0.2%      | 2.82559 +- 0.19331 | 0.90914 +- 0.15893 |  1.53958 +- 0.23118 |
|  pathways  |              narrower             |      0.2%      | 0.25554 +- 0.01072 | 0.21501 +- 0.01271 |  0.21512 +- 0.01217 |
|  pathways  | narrower.narrower*.(imports+narro |      0.2%      | 0.37245 

    Таблица:
| graph_name |               regex               | percent_starts |    time_tensor     |    time_bfs_ms     |     time_bfs_ss     |
|:----------:|:---------------------------------:|:--------------:|:------------------:|:------------------:|:-------------------:|
|    gzip    |            (Union_all)*           |      0.2%      | 0.06600 +- 0.00919 | 0.11013 +- 0.00951 |  0.11492 +- 0.00490 |
|    gzip    |                 a                 |      0.2%      | 0.05386 +- 0.00303 | 0.04389 +- 0.00089 |  0.04527 +- 0.00237 |
|    gzip    |             a.a*.(d+a)            |      0.2%      | 0.10893 +- 0.00453 | 0.10059 +- 0.00840 |  0.11142 +- 0.00589 |
|  pathways  |            (Union_all)*           |      0.2%      | 2.82559 +- 0.19331 | 0.90914 +- 0.15893 |  1.53958 +- 0.23118 |
|  pathways  |              narrower             |      0.2%      | 0.25554 +- 0.01072 | 0.21501 +- 0.01271 |  0.21512 +- 0.01217 |
|  pathways  | narrower.narrower*.(imports+narro |      0.2%      | 0.37245 +- 0.02318 | 0.27938 +- 0.04972 |  0.24649 +- 0.00976 |
|            |                wer)               |                |                    |                    |                     |
|    skos    |            (Union_all)*           |      1.0%      | 0.14070 +- 0.01037 | 0.14169 +- 0.00842 |  0.14016 +- 0.00681 |
|    skos    |                type               |      1.0%      | 0.01056 +- 0.00086 | 0.01086 +- 0.00120 |  0.01107 +- 0.00123 |
|    skos    | description.description*.(example |      1.0%      | 0.01271 +- 0.00085 | 0.01206 +- 0.00097 |  0.01214 +- 0.00097 |
|            |           +description)           |                |                    |                    |                     |
|  funding   |            (Union_all)*           |      1.0%      | 0.14249 +- 0.00752 | 0.16104 +- 0.00747 |  0.19728 +- 0.00722 |
|  funding   |                type               |      1.0%      | 0.02833 +- 0.00201 | 0.02751 +- 0.00234 |  0.02905 +- 0.00218 |
|  funding   | imports.imports*.(inverseOf+impor |      1.0%      | 0.03605 +- 0.00216 | 0.02457 +- 0.00155 |  0.02648 +- 0.00204 |
|            |                ts)                |                |                    |                    |                     |
|     wc     |            (Union_all)*           |      1.0%      | 0.01194 +- 0.00086 | 0.01523 +- 0.00096 |  0.01591 +- 0.00110 |
|     wc     |                 a                 |      1.0%      | 0.00965 +- 0.00069 | 0.01082 +- 0.00093 |  0.01116 +- 0.00086 |
|     wc     |             a.a*.(d+a)            |      1.0%      | 0.01798 +- 0.00129 | 0.01857 +- 0.00084 |  0.02038 +- 0.00170 |
|    bzip    |            (Union_all)*           |      1.0%      | 0.02055 +- 0.00174 | 0.03273 +- 0.00201 |  0.04246 +- 0.00297 |
|    bzip    |                 d                 |      1.0%      | 0.01804 +- 0.00188 | 0.01637 +- 0.00153 |  0.01754 +- 0.00135 |
|    bzip    |             a.a*.(d+a)            |      1.0%      | 0.03506 +- 0.00248 | 0.03674 +- 0.00312 |  0.04187 +- 0.00293 |
|    skos    |            (Union_all)*           |     10.0%      | 0.12948 +- 0.00647 | 0.21666 +- 0.01110 |  0.44559 +- 0.01066 |
|    skos    |                type               |     10.0%      | 0.01103 +- 0.00022 | 0.01338 +- 0.00014 |  0.01691 +- 0.00106 |
|    skos    | description.description*.(example |     10.0%      | 0.01240 +- 0.00079 | 0.01266 +- 0.00025 |  0.01620 +- 0.00083 |
|            |           +description)           |                |                    |                    |                     |
|  funding   |            (Union_all)*           |     10.0%      | 0.14312 +- 0.00650 | 0.21369 +- 0.00735 |  3.10677 +- 0.02751 |
|  funding   |                type               |     10.0%      | 0.02900 +- 0.00102 | 0.02964 +- 0.00137 |  0.08866 +- 0.00311 |
|  funding   | imports.imports*.(inverseOf+impor |     10.0%      | 0.03411 +- 0.00238 | 0.02863 +- 0.00173 |  0.08894 +- 0.00432 |
|            |                ts)                |                |                    |                    |                     |
|     wc     |            (Union_all)*           |     10.0%      | 0.01148 +- 0.00101 | 0.03272 +- 0.00257 |  0.20223 +- 0.00699 |
|     wc     |                 a                 |     10.0%      | 0.00983 +- 0.00078 | 0.01124 +- 0.00074 |  0.02869 +- 0.00223 |
|     wc     |             a.a*.(d+a)            |     10.0%      | 0.01766 +- 0.00179 | 0.01997 +- 0.00145 |  0.04948 +- 0.00305 |
|    bzip    |            (Union_all)*           |     10.0%      | 0.02109 +- 0.00153 | 0.04340 +- 0.00246 |  0.44453 +- 0.01325 |
|    bzip    |                 d                 |     10.0%      | 0.01786 +- 0.00014 | 0.01964 +- 0.00126 |  0.06338 +- 0.00295 |
|    bzip    |             a.a*.(d+a)            |     10.0%      | 0.03241 +- 0.00211 | 0.03831 +- 0.00191 |  0.13896 +- 0.00572 |
|    skos    |            (Union_all)*           |     30.0%      | 0.12747 +- 0.00802 | 0.21377 +- 0.00873 |  1.68088 +- 0.02331 |
|    skos    |                type               |     30.0%      | 0.01080 +- 0.00008 | 0.01400 +- 0.00063 |  0.03957 +- 0.00020 |
|    skos    | description.description*.(example |     30.0%      | 0.01254 +- 0.00087 | 0.01333 +- 0.00098 |  0.03069 +- 0.00156 |
|            |           +description)           |                |                    |                    |                     |
|  funding   |            (Union_all)*           |     30.0%      | 0.14159 +- 0.00956 | 0.23277 +- 0.00562 | 24.65210 +- 0.46489 |
|  funding   |                type               |     30.0%      | 0.02923 +- 0.00112 | 0.03428 +- 0.00182 |  0.42593 +- 0.00847 |
|  funding   | imports.imports*.(inverseOf+impor |     30.0%      | 0.03431 +- 0.00180 | 0.03361 +- 0.00158 |  0.39290 +- 0.00886 |
|            |                ts)                |                |                    |                    |                     |
|     wc     |            (Union_all)*           |     30.0%      | 0.01310 +- 0.00186 | 0.04215 +- 0.00216 |  1.11524 +- 0.01495 |
|     wc     |                 a                 |     30.0%      | 0.00967 +- 0.00061 | 0.01345 +- 0.00074 |  0.10328 +- 0.00279 |
|     wc     |             a.a*.(d+a)            |     30.0%      | 0.01758 +- 0.00097 | 0.02616 +- 0.00118 |  0.25405 +- 0.00476 |
|    bzip    |            (Union_all)*           |     30.0%      | 0.02245 +- 0.00146 | 0.05791 +- 0.00284 |  5.91616 +- 0.03567 |
|    bzip    |                 d                 |     30.0%      | 0.01792 +- 0.00091 | 0.02601 +- 0.00074 |  0.31796 +- 0.00783 |
|    bzip    |             a.a*.(d+a)            |     30.0%      | 0.03287 +- 0.00196 | 0.05360 +- 0.00272 |  0.86168 +- 0.01009 |

### По данным эксперимента можно сделать выводы:
1. Для маленьких графов таких как scos, wc, funding, bzip и даже gzip и при маленьком стартовом множестве (1.0 %), алгоритм номер 1 работает чуть бытрее или почти так же как алгоритм номер 2. 
2. Для большого графа, такого как pathways и при стартовом множестве 0.2 %, алгоритм номер 2 работает быстрее алгоритма 1. Разница была в 3-4 раза в сравнении с алгоритмом номер 1 в содержательном случае, то есть когда пути должны были быть и их должно быть много, а именно, когда regex = (Union all)*.
3. Для маленьких графов таких как scos, wc, funding, bzip и при стартовом множестве 10.0 и 30.0 %, алгоритм номер 1 работает бытрее алгоритма номер 2.

<b>Вывод:</b> алгоритм номер 2 работает быстрее алгоритма номер 1 при множестве от 6238 вершин, 12363 ребра и при стартовом множетсве 0.2 % или меньше от кол-ва вершин. При этом желательно, чтобы регулярный запрос был не пустым, чтобы время выполнения показывало не константу, а ассимптотику и как следствие алгоритм 2 был быстрее алгоритма 1.

# На сколько решение BfsMultipleSource быстрее решения BfsSingleSource при одинаковых начальных условиях ?

Берем разные графы по кол-ву вершин и процента стартовых вершин в них.

Эксперимент был выполнен до. Печатаем таблицу с данными эксперимента:

In [66]:
rows = []
for i in range(0, len(experiment_data), 3):
    rows.append([experiment_data[i][0], experiment_data[i][2], experiment_data[i][3], experiment_data[i+1][4], experiment_data[i+2][4]])
print_table(["graph_name", "regex", "percent_starts", "time_bfs_ms", "time_bfs_ss"], rows)

| graph_name |               regex               | percent_starts |    time_bfs_ms     |     time_bfs_ss     |
|:----------:|:---------------------------------:|:--------------:|:------------------:|:-------------------:|
|    gzip    |            (Union_all)*           |      0.2%      | 0.11013 +- 0.00951 |  0.11492 +- 0.00490 |
|    gzip    |                 a                 |      0.2%      | 0.04389 +- 0.00089 |  0.04527 +- 0.00237 |
|    gzip    |             a.a*.(d+a)            |      0.2%      | 0.10059 +- 0.00840 |  0.11142 +- 0.00589 |
|  pathways  |            (Union_all)*           |      0.2%      | 0.90914 +- 0.15893 |  1.53958 +- 0.23118 |
|  pathways  |              narrower             |      0.2%      | 0.21501 +- 0.01271 |  0.21512 +- 0.01217 |
|  pathways  | narrower.narrower*.(imports+narro |      0.2%      | 0.27938 +- 0.04972 |  0.24649 +- 0.00976 |
|            |                wer)               |                |                    |                     |
|

    Таблица:
| graph_name |               regex               | percent_starts |    time_bfs_ms     |     time_bfs_ss     |
|:----------:|:---------------------------------:|:--------------:|:------------------:|:-------------------:|
|    gzip    |            (Union_all)*           |      0.2%      | 0.11013 +- 0.00951 |  0.11492 +- 0.00490 |
|    gzip    |                 a                 |      0.2%      | 0.04389 +- 0.00089 |  0.04527 +- 0.00237 |
|    gzip    |             a.a*.(d+a)            |      0.2%      | 0.10059 +- 0.00840 |  0.11142 +- 0.00589 |
|  pathways  |            (Union_all)*           |      0.2%      | 0.90914 +- 0.15893 |  1.53958 +- 0.23118 |
|  pathways  |              narrower             |      0.2%      | 0.21501 +- 0.01271 |  0.21512 +- 0.01217 |
|  pathways  | narrower.narrower*.(imports+narro |      0.2%      | 0.27938 +- 0.04972 |  0.24649 +- 0.00976 |
|            |                wer)               |                |                    |                     |
|    skos    |            (Union_all)*           |      1.0%      | 0.14169 +- 0.00842 |  0.14016 +- 0.00681 |
|    skos    |                type               |      1.0%      | 0.01086 +- 0.00120 |  0.01107 +- 0.00123 |
|    skos    | description.description*.(example |      1.0%      | 0.01206 +- 0.00097 |  0.01214 +- 0.00097 |
|            |           +description)           |                |                    |                     |
|  funding   |            (Union_all)*           |      1.0%      | 0.16104 +- 0.00747 |  0.19728 +- 0.00722 |
|  funding   |                type               |      1.0%      | 0.02751 +- 0.00234 |  0.02905 +- 0.00218 |
|  funding   | imports.imports*.(inverseOf+impor |      1.0%      | 0.02457 +- 0.00155 |  0.02648 +- 0.00204 |
|            |                ts)                |                |                    |                     |
|     wc     |            (Union_all)*           |      1.0%      | 0.01523 +- 0.00096 |  0.01591 +- 0.00110 |
|     wc     |                 a                 |      1.0%      | 0.01082 +- 0.00093 |  0.01116 +- 0.00086 |
|     wc     |             a.a*.(d+a)            |      1.0%      | 0.01857 +- 0.00084 |  0.02038 +- 0.00170 |
|    bzip    |            (Union_all)*           |      1.0%      | 0.03273 +- 0.00201 |  0.04246 +- 0.00297 |
|    bzip    |                 d                 |      1.0%      | 0.01637 +- 0.00153 |  0.01754 +- 0.00135 |
|    bzip    |             a.a*.(d+a)            |      1.0%      | 0.03674 +- 0.00312 |  0.04187 +- 0.00293 |
|    skos    |            (Union_all)*           |     10.0%      | 0.21666 +- 0.01110 |  0.44559 +- 0.01066 |
|    skos    |                type               |     10.0%      | 0.01338 +- 0.00014 |  0.01691 +- 0.00106 |
|    skos    | description.description*.(example |     10.0%      | 0.01266 +- 0.00025 |  0.01620 +- 0.00083 |
|            |           +description)           |                |                    |                     |
|  funding   |            (Union_all)*           |     10.0%      | 0.21369 +- 0.00735 |  3.10677 +- 0.02751 |
|  funding   |                type               |     10.0%      | 0.02964 +- 0.00137 |  0.08866 +- 0.00311 |
|  funding   | imports.imports*.(inverseOf+impor |     10.0%      | 0.02863 +- 0.00173 |  0.08894 +- 0.00432 |
|            |                ts)                |                |                    |                     |
|     wc     |            (Union_all)*           |     10.0%      | 0.03272 +- 0.00257 |  0.20223 +- 0.00699 |
|     wc     |                 a                 |     10.0%      | 0.01124 +- 0.00074 |  0.02869 +- 0.00223 |
|     wc     |             a.a*.(d+a)            |     10.0%      | 0.01997 +- 0.00145 |  0.04948 +- 0.00305 |
|    bzip    |            (Union_all)*           |     10.0%      | 0.04340 +- 0.00246 |  0.44453 +- 0.01325 |
|    bzip    |                 d                 |     10.0%      | 0.01964 +- 0.00126 |  0.06338 +- 0.00295 |
|    bzip    |             a.a*.(d+a)            |     10.0%      | 0.03831 +- 0.00191 |  0.13896 +- 0.00572 |
|    skos    |            (Union_all)*           |     30.0%      | 0.21377 +- 0.00873 |  1.68088 +- 0.02331 |
|    skos    |                type               |     30.0%      | 0.01400 +- 0.00063 |  0.03957 +- 0.00020 |
|    skos    | description.description*.(example |     30.0%      | 0.01333 +- 0.00098 |  0.03069 +- 0.00156 |
|            |           +description)           |                |                    |                     |
|  funding   |            (Union_all)*           |     30.0%      | 0.23277 +- 0.00562 | 24.65210 +- 0.46489 |
|  funding   |                type               |     30.0%      | 0.03428 +- 0.00182 |  0.42593 +- 0.00847 |
|  funding   | imports.imports*.(inverseOf+impor |     30.0%      | 0.03361 +- 0.00158 |  0.39290 +- 0.00886 |
|            |                ts)                |                |                    |                     |
|     wc     |            (Union_all)*           |     30.0%      | 0.04215 +- 0.00216 |  1.11524 +- 0.01495 |
|     wc     |                 a                 |     30.0%      | 0.01345 +- 0.00074 |  0.10328 +- 0.00279 |
|     wc     |             a.a*.(d+a)            |     30.0%      | 0.02616 +- 0.00118 |  0.25405 +- 0.00476 |
|    bzip    |            (Union_all)*           |     30.0%      | 0.05791 +- 0.00284 |  5.91616 +- 0.03567 |
|    bzip    |                 d                 |     30.0%      | 0.02601 +- 0.00074 |  0.31796 +- 0.00783 |
|    bzip    |             a.a*.(d+a)            |     30.0%      | 0.05360 +- 0.00272 |  0.86168 +- 0.01009 |

### По данным эксперимента можно сделать выводы:
1. Для маленьких графов таких как scos, wc, funding, bzip и даже gzip и при маленьком стартовом множестве (1.0 %), алгоритм BfsMultipleSource работает чуть бытрее или почти так же как BfsSingleSource. 
2. Для большого графа, такого как pathways и при стартовом множестве 0.2 %, BfsMultipleSource работает быстрее BfsSingleSource. Разница была на 70% в сравнении с BfsSingleSource в содержательном случае, то есть когда пути должны были быть и их должно быть много, а именно, когда regex = (Union all)*.
3. Для маленьких графов таких как scos, wc, funding, bzip и при стартовом множестве 10.0 %, алгоритм BfsMultipleSource работает бытрее алгоритма BfsSingleSource от 2 до 15 раз для regex = (Union all)*. Для менее содержательного regex BfsMultipleSource работает бытрее BfsSingleSource от 0.25 до 3.5 раз.
4. Для маленьких графов таких как scos, wc, funding, bzip и при стартовом множестве 30.0 %, алгоритм BfsMultipleSource работает бытрее алгоритма BfsSingleSource от 8 до 105 раз для regex = (Union all)*. Для менее содержательного regex BfsMultipleSource работает бытрее BfsSingleSource от 2 до 12.5 раз.

<b>Вывод: </b>разница в скорости между BfsMultipleSource и BfsSingleSource экспоненциально зависит от кол-ва стартовых вершин.



# Заключение: 

Так как основная вычислительная сложность алгоритмов основана на умножении матриц, то нет разницы какую использовать матрицу [csr](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix) и [csc](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix) из [scipy.sparse](https://docs.scipy.org/doc/scipy/reference/sparse.html) для алгоритмов.

Алгоритм номер 2 работает быстрее алгоритма номер 1 при множестве от 6238 вершин, 12363 ребра и при стартовом множетсве 0.2 % или меньше от кол-ва вершин. При этом желательно, чтобы регулярный запрос был не пустым, чтобы время выполнения показывало не константу, а ассимптотику и как следствие алгоритм 2 был быстрее алгоритма 1.

Разница в скорости между BfsMultipleSource и BfsSingleSource экспоненциально зависит от кол-ва стартовых вершин.